### PR TITLE
AB#43 allow spectating

### PR DESCRIPTION
**Pull request by michaelschufi**
This is more work than expected...

Implementation of the spectating feature. I did:
- Added RoleRequest and RoleResponse messages. 
- Updated MainController and GameController to handle the messages.
- Added the role switch feature in the frontend:
   - Eliminated Dialogs :)
   - Added Welcome message when no role has been selected
   - Added the buttons to switch roles

TODO:

- [x] Cleanup Game component
- [x] Rewrite tests 😢 
- [x] General cleanup

### CODE REVIEW

**Comment by sir-marc** [comment]
There seem to be some issues:
Scenario:
Open the game with two players, then once the game is started switch between spectating and playing mode back and forth. 

Is: You lose the game 
Should: nothing should happen until the next game starts

Scenario 2:
Open the game with two players.
Then reload the page (for one player) and select "spectating" (this selection should not appear, as the user is already playing). Then the unit flashes green and red.

**Comment by michaelschufi** [comment]
I have fixed both scenarios. Could you have a look?

File: game-frontend/src/Game.tsx [hunk 1]
```diff
@@ -26,27 +27,47 @@ function Game(): JSX.Element {
...
+    isPlayer && matchIsRunning && !!gameState.objectMap[userContext.unitId];
+  const isOnlyOnePlayerLeft =
+    gameState.objectIds
+      .map(id => gameState.objectMap[id])
+      .filter(gameObject => gameObject.type === GameObjectTypes.Unit).length ===
+    1;
   const isDead = isPlayer && hasJoinedGame && !isLiving;
+  const hasWon = isPlayer && hasJoinedGame && isLiving && isOnlyOnePlayerLeft;
+
+  console.log({
```
Comment (StuckiSimon) [inline game-frontend/src/Game.tsx]: Not quite sure how ready for review this PR is? :)

**Comment by sir-marc** [comment]
> I have fixed both scenarios. Could you have a look?

As both issues relied on switching roles during a match and because that is no longer possible, the issues are no longer a problem 👍

File: game-frontend/src/Game.tsx (hunk 1 above)
Comment (michaelschufi) [reply game-frontend/src/Game.tsx]: I wanted @sir-marc to have a look at the front-end before finishing and cleaning it up :)

Comment (michaelschufi) [reply game-frontend/src/Game.tsx]: Btw, how can you set the PR as a draft?

Comment (StuckiSimon) [reply game-frontend/src/Game.tsx]: Upon creation you can choose whether to create a PR or create a draft PR

Comment (michaelschufi) [reply game-frontend/src/Game.tsx]: There's a small link below the assigned reviewers. Didn't catch that before ^^

**Comment by michaelschufi** [comment]
Finally done with adding/updating the tests.

Edit: Not so sure about the implementation regarding persisting the role choice on the frontend. It surely could be cleaner.

File: game-backend/src/GameController.ts [hunk 2]
```diff
@@ -23,7 +23,7 @@ interface TickFeedback {
 }
 
 export default class GameController {
-  private isRunning = false;
+  public isRunning = false;
```
Comment (StuckiSimon) [inline game-backend/src/GameController.ts]: Why public?

File: game-frontend/src/Game.tsx [hunk 3]
```diff
@@ -23,26 +26,38 @@ function Game(): JSX.Element {
...
   const userContext = useUserContext();
-  const playStartSound = useStartSound();
+  const role = useRole();
 
-  const isLiving = !!gameState.objectMap[userContext.unitId];
+  const matchIsRunning =
+    gameState.objectIds
+      .map(id => gameState.objectMap[id])
+      .filter(gameObject => gameObject.type === GameObjectTypes.Unit).length >
+    0;
```
Comment (StuckiSimon) [inline game-frontend/src/Game.tsx]: Why not use `some()`?

File: game-frontend/src/Game.tsx [hunk 4]
```diff
@@ -98,11 +124,12 @@ function Game(): JSX.Element {
             <GameObjects
               objectIds={gameState.objectIds}
               objectMap={gameState.objectMap}
-              ownUnitId={userContext.unitId}
+              ownUnitId={userContext ? userContext.unitId : ''}
             />
           </Container>
         </Stage>
       </CanvasWrapper>
+      {!matchIsRunning ? <RoleSwitch /> : ''}
```
Comment (StuckiSimon) [inline game-frontend/src/Game.tsx]: why not render `null` instead of `''`?

File: game-frontend/src/components/role-switch/RoleSwitchButton.tsx [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+interface Props {
+  roleToRequest: Role;
+}
+
+const RoleSwitchButton: FC<Props> = ({ roleToRequest, children }) => {
+  const sendMessage = useSendMessage();
+
+  const sendRoleRequest = useMemo(
+    () => (newRole: Role): void => {
```
Comment (StuckiSimon) [inline game-frontend/src/components/role-switch/RoleSwitchButton.tsx]: Why not `useCallback`?

Comment (StuckiSimon) [reply game-frontend/src/components/role-switch/RoleSwitchButton.tsx]: Furthermore, what's the benefit of memoing this? The onClick event listener is re-created upon each rerender anyway.

File: game-frontend/src/components/role-switch/index.tsx [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+  const role = useRole();
+
+  const roleToRequest = role === Role.SPECTATOR ? Role.PLAYER : Role.SPECTATOR;
+
+  return role ? (
+    <RoleSwitchButton roleToRequest={roleToRequest}>
+      Ich möchte {getRoleText(roleToRequest)}!
+    </RoleSwitchButton>
+  ) : (
+    <></>
```
Comment (StuckiSimon) [inline game-frontend/src/components/role-switch/index.tsx]: Why not `null`?

File: game-frontend/src/service/game/context.ts [hunk 7]
```diff
@@ -11,14 +11,17 @@ import {
...
 const useUserId = createPersistedState(USER_ID_QUERY_PARAM);
+const usePersistRole = createPersistedState(ROLE_LOCAL_STORAGE_KEY);
 
 export const WebsocketContext = createContext<ContextState>(
   emptyWebsocketContext
 );
 
 export function useWebsocket(): ContextState {
   const [userId, setUserId] = useUserId<UserId>(() => uuidv4());
+  const [role, setRole] = usePersistRole<Role | null>(() => null);
```
Comment (StuckiSimon) [inline game-frontend/src/service/game/context.ts]: The callback for `usePersistRole` is not required, simply do `usePersistRole<Role | null>(null);` in this case.

File: game-frontend/src/service/game/context.ts [hunk 8]
```diff
@@ -34,11 +37,30 @@ export function useWebsocket(): ContextState {
   useEffect(() => {
     if (socket) {
       socket.addEventListener('message', event => {
-        const action: Message = JSON.parse(event.data);
-        dispatch(action);
+        const message: Message = JSON.parse(event.data);
+        dispatch(message);
```
Comment (StuckiSimon) [inline game-frontend/src/service/game/context.ts]: This was actually called `action` on purpose.

File: game-frontend/src/service/game/context.ts [hunk 9]
```diff
@@ -34,11 +37,30 @@ export function useWebsocket(): ContextState {
...
       });
+      // persist role choice
+      socket.addEventListener('message', event => {
```
Comment (StuckiSimon) [inline game-frontend/src/service/game/context.ts]: No need to add a different event listener. This should be handled within the other event listener.

File: game-frontend/src/service/game/context.ts [hunk 10]
```diff
@@ -34,11 +37,30 @@ export function useWebsocket(): ContextState {
...
+        if (message.type === 'roleResponse') {
+          const { newRole } = (message as RoleResponse).payload;
+          setRole(newRole);
+        }
+      });
+    }
+  }, [socket, setUserId, dispatch, setRole]);
+
+  // set initial role from localStorage
+  useEffect(() => {
```
Comment (StuckiSimon) [inline game-frontend/src/service/game/context.ts]: Not really the nicest way to handle this :)

File: game-frontend/src/service/game/context.ts [hunk 11]
```diff
@@ -34,11 +37,30 @@ export function useWebsocket(): ContextState {
...
-        dispatch(action);
+        const message: Message = JSON.parse(event.data);
+        dispatch(message);
       });
+      // persist role choice
+      socket.addEventListener('message', event => {
+        const message: Message = JSON.parse(event.data);
+        if (message.type === 'roleResponse') {
+          const { newRole } = (message as RoleResponse).payload;
+          setRole(newRole);
```
Comment (StuckiSimon) [inline game-frontend/src/service/game/context.ts]: Storing the role twice kind of defeats the purpose of a `single source of truth`.

**Review by StuckiSimon — COMMENTED** [review]
Had a quick look over the implementation - didn't thoroughly test it.

File: game-frontend/src/service/game/context.ts (hunk 11 above)
Comment (StuckiSimon) [reply game-frontend/src/service/game/context.ts]: There is actually no need to store it within the other context - it's quite easy to provide the localstorage state (`usePersistRole`) directly for other components to use.

File: game-backend/src/GameController.ts (hunk 2 above)
Comment (michaelschufi) [reply game-backend/src/GameController.ts]: Forget to change that back ^^, thanks!

**Comment by michaelschufi** [comment]
With the help of @StuckiSimon we are not using a parallel store anymore.

Also: I fixed a bug in the game controller, which allowed adding an user multiple times.

**Comment by mailaenderli** [comment]
I found three points:
- The sounds (Starting game and loosing) are playing if you switch between the roles while a game is running
- the game starts despite only one is playing. Steps to reproduce:
  - both are spectating
  - one clicks play and then spectate
  - wait a few secs
  - the other clicks on play
- your own unit "blinks" red-green if you switch from player to spectator in a game (you need to switch two times to see the effect)

**Comment by michaelschufi** [comment]
> * the game starts despite only one is playing. Steps to reproduce:
>    * both are spectating
>    * one clicks play and then spectate
>    * wait a few secs
>    * the other clicks on play

I can reproduce that and will work on that.

> * The sounds (Starting game and loosing) are playing if you switch between the roles while a game is running
and
> * your own unit "blinks" red-green if you switch from player to spectator in a game (you need to switch two times to see the effect)

The role switch button shouldn't show up while the game is running. I'm a bit confused 😅

**Comment by mailaenderli** [comment]
> > * The sounds (Starting game and loosing) are playing if you switch between the roles while a game is running
> >   and
> > * your own unit "blinks" red-green if you switch from player to spectator in a game (you need to switch two times to see the effect)
> 
> The role switch button shouldn't show up while the game is running. I'm a bit confused 😅  

Yep that's right. I wasn't somhow working on the newest commit 🙃

**Comment by michaelschufi** [comment]
old version of #112  and no longer needed

**Comment by StuckiSimon** [comment]
Closed in favor of #112

**Pull request closed without merging**